### PR TITLE
Added support to create asymmetrical U-Net structures

### PR DESCRIPTION
**Pull request by Gothos**
What does this PR do? 

1.  Adds ```UNetMidBlock2D``` as a supported ```mid_block_type``` in the ```UNet2DConditionModel``` class.
2. Adds basic support for specifying ```transformer_layers_per_block``` for every layer of the ```UNet2DConditionModel``` class.
Support is somewhat rigid at the moment, using this functionality requires specifying the ```reverse_transformer_layers_per_block``` attribute in a config (set to ```None``` by default). 
Please comment on any style changes/bugs you find in the code, I'll fix it as soon as I can.

@sayakpaul

### CODE REVIEW

**Review by sayakpaul — COMMENTED** [review]
Thanks for working on this. The changes seem clean and simple to me. 
@patrickvonplaten WDYT?

**Review by patrickvonplaten — COMMENTED** [review]
Hmm I would need to see a clear use case for when asymmetrical U-Nets make sense. I'm a bit hesitant to adapt the Unet2DCondition class for use case that have not been verified to work yet.

Do we have a checkpoint that shows good results with an asymmetrical U-Net?

**Comment by Gothos** [comment]
The checkpoint we're using for the SDXL distillation is asymmetrical. We did put those results up on slack.

**Comment by sayakpaul** [comment]
Hi @Gothos!

You can check out this test suite to get an idea:
https://github.com/huggingface/diffusers/blob/main/tests/models/test_models_unet_2d_condition.py

I'd instantiate an asymmetric UNet in that test suite and run dummy inference and perform assertions as you'd see for some of the tests being done in the test suite mentioned above.

Let me know if that makes sense?

**Comment by Gothos** [comment]
Yes, thanks! I'll get this done as quick as I can.

**Comment by Gothos** [comment]
@sayakpaul  I've written a basic test asserting input shape equals the output shape for an asymmetrical UNet. Please let me know if anything's wrong/ any other tests are required.

File: src/diffusers/models/unet_2d_blocks.py [hunk 1]
```diff
@@ -624,6 +624,11 @@ def __init__(
         self.num_attention_heads = num_attention_heads
         resnet_groups = resnet_groups if resnet_groups is not None else min(in_channels // 4, 32)
 
+        # support for variable transformer layers per block 
+
```
Comment (sayakpaul) [inline src/diffusers/models/unet_2d_blocks.py:628]:
```suggestion
```

File: src/diffusers/models/unet_2d_condition.py [hunk 2]
```diff
@@ -105,10 +106,14 @@ class UNet2DConditionModel(ModelMixin, ConfigMixin, UNet2DConditionLoadersMixin)
...
             The dimension of the cross attention features.
-        transformer_layers_per_block (`int` or `Tuple[int]`, *optional*, defaults to 1):
+        transformer_layers_per_block (`int`, `Tuple[int]`, or Tuple[Tuple] , *optional*, defaults to 1):
             The number of transformer blocks of type [`~models.attention.BasicTransformerBlock`]. Only relevant for
             [`~models.unet_2d_blocks.CrossAttnDownBlock2D`], [`~models.unet_2d_blocks.CrossAttnUpBlock2D`],
             [`~models.unet_2d_blocks.UNetMidBlock2DCrossAttn`].
+        reverse_transformer_layers_per_block : (`Tuple[Tuple]`. *optional*, required if passing transformer_layers_per_block of type `Tuple[Tuple]):
+        Provides the same functionality as transformer_layer_per_block, but for the up_blocks in the U-Net. Only relevant for
+            [`~models.unet_2d_blocks.CrossAttnDownBlock2D`], [`~models.unet_2d_blocks.CrossAttnUpBlock2D`],
+            [`~models.unet_2d_blocks.UNetMidBlock2DCrossAttn`].
```
Comment (sayakpaul) [inline src/diffusers/models/unet_2d_condition.py:116]: I don't think you're following right docstring format here. Please refer to how other legacy args are documented.

File: src/diffusers/models/unet_2d_condition.py [hunk 3]
```diff
@@ -419,6 +431,8 @@ def __init__(
         if isinstance(layers_per_block, int):
             layers_per_block = [layers_per_block] * len(down_block_types)
 
+
+
```
Comment (sayakpaul) [inline src/diffusers/models/unet_2d_condition.py:435]: Can remove one.

File: tests/models/test_models_unet_2d_condition.py [hunk 4]
```diff
@@ -606,6 +606,22 @@ def test_pickle(self):
...
+
+        torch.manual_seed(0)
+        model = self.model_class(**init_dict)
+        model.to(torch_device)
+
+        output = model(**inputs_dict).sample
+        expected_shape = inputs_dict["sample"].shape
+
+        # Check if input and output shapes are the same
+        self.assertEqual(output.shape, expected_shape , "Input and output shapes do not match")
```
Comment (sayakpaul) [inline tests/models/test_models_unet_2d_condition.py:623]: Works for me.

**Review by sayakpaul — APPROVED** [review]
Good to merge for me. Thanks for the contribution!

Let's make sure to resolve the rest of the comments and ensure the CI is green.

File: src/diffusers/models/unet_2d_condition.py (hunk 3 above)
Comment (Gothos) [reply src/diffusers/models/unet_2d_condition.py]: Followed your suggestion.

File: src/diffusers/models/unet_2d_condition.py (hunk 2 above)
Comment (Gothos) [reply src/diffusers/models/unet_2d_condition.py:116]: Would this do? I'm not 100% sure of my style here.

**Comment by Gothos** [comment]
Most of the errors seem to step from the addition of `UNetMidBlock2D` as a supported type. I'm trying to figure out where the problem is.

File: src/diffusers/models/unet_2d_condition.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/models/unet_2d_condition.py:116]: Keep the modifications to `reverse_transformer_layers_per_block` only and your proposed phrase works for me.

Comment (Gothos) [reply src/diffusers/models/unet_2d_condition.py:116]: All right, thanks!

**Comment by Gothos** [comment]
I think I've fixed the bug. It should work now.

**Comment by sayakpaul** [comment]
@Gothos could you run `make style && make quality` on your end? 

Also `make fix-copies` as well. Without that, we cannot merge the PR.

**Comment by Gothos** [comment]
All right, I'll do that in a bit. Thanks for the help!

**Comment by Gothos** [comment]
@sayakpaul  I've done the needful.

**Comment by Gothos** [comment]
Running the commands above to enforce consistency seems to have attempted to make "UNetMidBlockFlat" a viable mid block type under the versatile diffusion pipeline in `modeling_text_unet.py`. However this class is not implemented in the file and this causes an error.

**Comment by sayakpaul** [comment]
We need to enforce repository consistency, sadly. 

Is it possible to copy over the implementation in `modeling_text_unet.py`?

**Comment by Gothos** [comment]
> We need to enforce repository consistency, sadly.
> 
> Is it possible to copy over the implementation in `modeling_text_unet.py`?

Sure! I've just done it and rerun the makes twice to make sure nothing popped up.

**Review by sayakpaul — APPROVED** [review]
Good work!

File: src/diffusers/models/unet_2d_blocks.py [hunk 5]
```diff
@@ -587,7 +587,7 @@ def __init__(
         self.attentions = nn.ModuleList(attentions)
         self.resnets = nn.ModuleList(resnets)
 
-    def forward(self, hidden_states, temb=None):
+    def forward(self, hidden_states, temb=None, *args, **kwargs):
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:590]:
```suggestion
    def forward(self, hidden_states, temb=None):
```
Is this really needed?

File: tests/models/test_models_unet_2d_condition.py [hunk 6]
```diff
@@ -606,6 +606,22 @@ def test_pickle(self):
 
         assert (sample - sample_copy).abs().max() < 1e-4
 
+    def test_asymmetrical_unet(self):
```
Comment (patrickvonplaten) [inline tests/models/test_models_unet_2d_condition.py:609]: nice test!

File: src/diffusers/models/unet_2d_blocks.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/diffusers/models/unet_2d_blocks.py:590]: Can we remove this, I don't see exactly how this is required here

File: tests/models/test_models_unet_2d_condition.py (hunk 6 above)
Comment (Gothos) [reply tests/models/test_models_unet_2d_condition.py:609]:
``` # 4. mid
        if self.mid_block is not None:
            sample = self.mid_block(
                sample,
                emb,
                encoder_hidden_states=encoder_hidden_states,
                attention_mask=attention_mask,
                cross_attention_kwargs=cross_attention_kwargs,
                encoder_attention_mask=encoder_attention_mask,
            )```
            These lines make the *args, **kwargs necessary. Unless I should add a the ``` has_attention``` attribute and work off of that, like the Downsampling and upsampling blocks?

File: src/diffusers/models/unet_2d_blocks.py (hunk 5 above)
Comment (Gothos) [reply src/diffusers/models/unet_2d_blocks.py:590]:
``` # 4. mid
        if self.mid_block is not None:
            sample = self.mid_block(
                sample,
                emb,
                encoder_hidden_states=encoder_hidden_states,
                attention_mask=attention_mask,
                cross_attention_kwargs=cross_attention_kwargs,
                encoder_attention_mask=encoder_attention_mask,
            )```
         
 These lines make the *args, **kwargs necessary. Unless I should add a ``` has_attention``` attribute and work off of that, like the Downsampling and upsampling blocks?

Comment (Gothos) [reply src/diffusers/models/unet_2d_blocks.py:590]:
```suggestion
            if self.mid_block is not None:
            if hasattr(self.mid_block, "has_cross_attention") and self.mid_block.has_cross_attention:
                sample = self.mid_block(
                    sample,
                    emb,
                    encoder_hidden_states=encoder_hidden_states,
                    attention_mask=attention_mask,
                    cross_attention_kwargs=cross_attention_kwargs,
                    encoder_attention_mask=encoder_attention_mask,
                )
            else:
                sample = self.mid_block(
                    sample,
                    emb,
                )
```
Would following your suggestion above and then doing this work, @patrickvonplaten?

Comment (sayakpaul) [reply src/diffusers/models/unet_2d_blocks.py:590]: This sounds good to me!

Comment (Gothos) [reply src/diffusers/models/unet_2d_blocks.py:620]: All right, I'll commit the changes. Just wanted to check once.

**Comment by sayakpaul** [comment]
@patrickvonplaten could you give this a look too?

File: src/diffusers/models/unet_2d_condition.py [hunk 7]
```diff
@@ -1062,14 +1090,21 @@ def forward(
...
+                    encoder_hidden_states=encoder_hidden_states,
+                    attention_mask=attention_mask,
+                    cross_attention_kwargs=cross_attention_kwargs,
+                    encoder_attention_mask=encoder_attention_mask,
+                )
+            else:
+                sample = self.mid_block(
+                    sample,
+                    emb,
+                )
```
Comment (sayakpaul) [inline src/diffusers/models/unet_2d_condition.py:1106]: nit: This is can wrapped into a single line:

```suggestion
                sample = self.mid_block(sample, emb)
```

**Review by sayakpaul — APPROVED** [review]
Thanks a lot for working on this! PR is good to go from my end.

**Pull request merged**